### PR TITLE
SM client implementation

### PR DESCRIPTION
**Pull request by Eabramov84**

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -60,6 +60,7 @@ type Migration struct {
 
 // Config instance.
 type Config struct {
+	NodeID             string `json:"nodeId"`
```
Comment (al1img) [inline config/config.go:63]: Let's take from IAM?

File: smclient/smclient.go [hunk 2]
```diff
@@ -0,0 +1,865 @@
...
+	}
+}
+
+func (client *SMClient) processGetUnitConfigStatus() {
+	status := &pb.UnitConfigStatus{VendorVersion: client.unitConfigProcessor.GetUnitConfigInfo()}
+
+	if err := client.stream.Send(&pb.SMOutgoingMessages{
+		SMOutgoingMessage: &pb.SMOutgoingMessages_UnitConfigStatus{UnitConfigStatus: status},
+	}); err != nil {
+		log.Error("Can't send unit config status")
```
Comment (al1img) [inline smclient/smclient.go:338]: %v, err

File: smclient/smclient.go [hunk 3]
```diff
@@ -0,0 +1,865 @@
...
+}
+
+func (client *SMClient) processNodeMonitoringData() {
+	if err := client.stream.Send(
+		&pb.SMOutgoingMessages{
+			SMOutgoingMessage: &pb.SMOutgoingMessages_NodeMonitoring{
+				NodeMonitoring: cloudprotocolMonitoringToPB(client.monitoringProvider.GetNodeMonitoringData()),
+			},
+		}); err != nil {
+		log.Errorf("Can't send monitoring notification: %s ", err)
```
Comment (al1img) [inline smclient/smclient.go:509]: %v

File: smclient/smclient.go [hunk 4]
```diff
@@ -0,0 +1,865 @@
...
+	}
+}
+
+func (client *SMClient) handleChannels() {
+	for {
+		select {
+		case runtimeStatus := <-client.runtimeStatusChannel:
+			if err := client.sendRuntimeInstanceNotifications(runtimeStatus); err != nil {
+				log.Errorf("Can't send runtime instance notification: %s", err)
```
Comment (al1img) [inline smclient/smclient.go:518]: %v

File: smclient/smclient.go [hunk 5]
```diff
@@ -0,0 +1,865 @@
...
+			if err != nil {
+				log.Errorf("Can't convert alert to pb: %s", err)
+
+				continue
+			}
+
+			alertNtf := &pb.SMOutgoingMessages_Alert{Alert: pbAlert}
+
+			if err := client.stream.Send(&pb.SMOutgoingMessages{SMOutgoingMessage: alertNtf}); err != nil {
+				log.Errorf("Can't send alert: %s ", err)
```
Comment (al1img) [inline smclient/smclient.go:534]: %v, remove extra space

File: smclient/smclient.go [hunk 6]
```diff
@@ -0,0 +1,865 @@
...
+			}
+
+		case monitoringData := <-client.monitoringChannel:
+			if err := client.stream.Send(
+				&pb.SMOutgoingMessages{
+					SMOutgoingMessage: &pb.SMOutgoingMessages_NodeMonitoring{
+						NodeMonitoring: cloudprotocolMonitoringToPB(monitoringData),
+					},
+				}); err != nil {
+				log.Errorf("Can't send monitoring notification: %s ", err)
```
Comment (al1img) [inline smclient/smclient.go:546]: %v, remove extra space

File: smclient/smclient.go [hunk 7]
```diff
@@ -0,0 +1,865 @@
...
+
+		case stateRquest := <-client.stateReqChannel:
+			if err := client.stream.Send(
+				&pb.SMOutgoingMessages{SMOutgoingMessage: &pb.SMOutgoingMessages_InstanceStateRequest{
+					InstanceStateRequest: &pb.InstanceStateRequest{
+						Instance: pbconvert.InstanceIdentToPB(stateRquest.InstanceIdent),
+						Default:  stateRquest.Default,
+					},
+				}}); err != nil {
+				log.Errorf("Can't send state request notification :%s ", err)
```
Comment (al1img) [inline smclient/smclient.go:572]: %v, remove extra space

File: smclient/smclient.go [hunk 8]
```diff
@@ -0,0 +1,865 @@
...
+
+				return
+			}
+
+		case logs := <-client.logsChannel:
+			if err := client.stream.Send(
+				&pb.SMOutgoingMessages{
+					SMOutgoingMessage: &pb.SMOutgoingMessages_Log{Log: cloudprotocolLogToPB(logs)},
+				}); err != nil {
+				log.Errorf("Can't send logs: %s ", err)
```
Comment (al1img) [inline smclient/smclient.go:582]: %v, remove extra space

File: smclient/smclient.go [hunk 9]
```diff
@@ -0,0 +1,865 @@
...
+}
+
+func (client *SMClient) sendRuntimeInstanceNotifications(runtimeStatus launcher.RuntimeStatus) error {
+	if runtimeStatus.RunStatus != nil {
+		runStatusNtf := &pb.SMOutgoingMessages_RunInstancesStatus{
+			RunInstancesStatus: runInstanceStatusToPB(runtimeStatus.RunStatus),
+		}
+
+		if err := client.stream.Send(&pb.SMOutgoingMessages{SMOutgoingMessage: runStatusNtf}); err != nil {
+			return aoserrors.Errorf("Can't send runtime status notification: %s ", err)
```
Comment (al1img) [inline smclient/smclient.go:600]: %v, remove extra space

File: smclient/smclient.go [hunk 10]
```diff
@@ -0,0 +1,865 @@
...
+		}
+	}
+
+	if runtimeStatus.UpdateStatus != nil {
+		updateStatusNtf := &pb.SMOutgoingMessages_UpdateInstancesStatus{
+			UpdateInstancesStatus: updateInstanceStatusToPB(runtimeStatus.UpdateStatus),
+		}
+
+		if err := client.stream.Send(&pb.SMOutgoingMessages{SMOutgoingMessage: updateStatusNtf}); err != nil {
+			return aoserrors.Errorf("Can't send update status notification: %s ", err)
```
Comment (al1img) [inline smclient/smclient.go:610]: %v, remove extra space

File: smclient/smclient_test.go [hunk 11]
```diff
@@ -0,0 +1,795 @@
...
+	"os"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/aoscloud/aos_common/aoserrors"
+	"github.com/aoscloud/aos_common/aostypes"
+	"github.com/aoscloud/aos_common/api/cloudprotocol"
+	pb "github.com/aoscloud/aos_common/api/servicemanager/v3"
+	"github.com/aoscloud/aos_servicemanager/config"
```
Comment (al1img) [inline smclient/smclient_test.go:33]: local package

File: smclient/smclient_test.go [hunk 12]
```diff
@@ -0,0 +1,795 @@
...
+	if err != nil {
+		t.Fatalf("Can't create test server: %s", err)
+	}
+
+	defer server.close()
+
+	client, err := smclient.New(&config.Config{CMServerURL: serverURL, NodeID: "mainSM"}, nil, nil, nil, nil, nil,
+		nil, nil, nil, true)
+	if err != nil {
+		t.Fatalf("Can't create UM client: %s", err)
```
Comment (al1img) [inline smclient/smclient_test.go:122]: %v

File: smclient/smclient_test.go [hunk 13]
```diff
@@ -0,0 +1,795 @@
...
+	}
+	defer client.Close()
+
+	if err = server.waitClientRegistered(); err != nil {
+		t.Fatalf("Sm registration error: %s", err)
```
Comment (al1img) [inline smclient/smclient_test.go:127]: %v

File: smclient/smclient_test.go [hunk 14]
```diff
@@ -0,0 +1,795 @@
...
+}
+
+func TestInstanceMessages(t *testing.T) {
+	// TODO: implement tests
+}
+
+func TestInstanceStateProcessing(t *testing.T) {
+	server, err := newTestServer(serverURL)
+	if err != nil {
+		t.Fatalf("Can't create test server: %s", err)
```
Comment (al1img) [inline smclient/smclient_test.go:138]: %v for err in all cases

File: smclient/smclient_test.go [hunk 15]
```diff
@@ -0,0 +1,795 @@
...
+
+	client, err := smclient.New(&config.Config{CMServerURL: serverURL, NodeID: "mainSM"}, &stateHandler, nil, nil, nil,
+		nil, nil, nil, nil, true)
+	if err != nil {
+		t.Fatalf("Can't create UM client: %s", err)
+	}
+	defer client.Close()
+
+	if err = server.waitClientRegistered(); err != nil {
+		t.Fatalf("Sm registration error: %s", err)
```
Comment (al1img) [inline smclient/smclient_test.go:156]: Use SM abbreviation for service manager

File: smclient/smclient_test.go [hunk 16]
```diff
@@ -0,0 +1,795 @@
...
+
+	stateHandler.newStateChan <- cloudprotocol.NewState{
+		InstanceIdent: aostypes.InstanceIdent{ServiceID: "service1", SubjectID: "s1", Instance: 2},
+		Checksum:      checkSumStr, State: stateStr,
+	}
+
+	receivedNewState := <-server.newInstanceStateChannel
+
+	if !proto.Equal(expectedNewState, receivedNewState.NewInstanceState) {
+		t.Error("received new state doesn't match sent data")
```
Comment (al1img) [inline smclient/smclient_test.go:178]: Capital letter

File: config/config.go (hunk 1 above)
Comment (Eabramov84) [reply config/config.go:63]: commit was removed

**Review by MykolaSolyanko — APPROVED** [review]

**Review by MykolaSolyanko — APPROVED** [review]

File: smclient/smclient.go [hunk 17]
```diff
@@ -0,0 +1,822 @@
...
+// UnitConfigProcessor unit configuration handler.
+type UnitConfigProcessor interface {
+	GetUnitConfigInfo() (version string)
+	CheckUnitConfig(configJSON, version string) error
+	UpdateUnitConfig(configJSON, version string) error
+}
+
+// ServicesProcessor process desired services list.
+type ServicesProcessor interface {
+	ProcesdDesiredServices(services []aostypes.ServiceInfo) error
```
Comment (al1img) [inline smclient/smclient.go:95]: Process

File: smclient/smclient.go [hunk 18]
```diff
@@ -0,0 +1,822 @@
...
+}
+
+// LayersProcessor process desired layer list.
+type LayersProcessor interface {
+	ProcessDesiredLayers(layers []aostypes.LayerInfo) error
+}
+
+// InstanceLauncher service instances launcher interface.
+type InstanceLauncher interface {
+	RunInstances(instances []aostypes.InstanceInfo)
```
Comment (al1img) [inline smclient/smclient.go:105]: Add `forceReset` param

File: smclient/smclient.go [hunk 19]
```diff
@@ -0,0 +1,822 @@
...
+// Close closes SM client.
+func (client *SMClient) Close() (err error) {
+	log.Debug("Close SM client")
+
+	if client.stream != nil {
+		err = aoserrors.Wrap(client.stream.CloseSend())
+	}
+
+	if client.connection != nil {
+		client.connection.Close()
```
Comment (al1img) [inline smclient/smclient.go:178]: check error

File: smclient/smclient.go [hunk 20]
```diff
@@ -0,0 +1,822 @@
...
+	}
+
+	close(client.closeChannel)
+
+	return aoserrors.Wrap(err)
```
Comment (al1img) [inline smclient/smclient.go:183]: Already wrapped

File: smclient/smclient.go [hunk 21]
```diff
@@ -0,0 +1,822 @@
...
+	if err := client.stream.Send(&pb.SMOutgoingMessages{
+		SMOutgoingMessage: &pb.SMOutgoingMessages_UnitConfigStatus{UnitConfigStatus: status},
+	}); err != nil {
+		log.Errorf("Can't send unit config status: %v", err)
+	}
+}
+
+func (client *SMClient) processCheckUnitConfig(check *pb.CheckUnitConfig) {
+	status := &pb.UnitConfigStatus{}
+	if err := client.unitConfigProcessor.CheckUnitConfig(check.UnitConfig, check.VendorVersion); err != nil {
```
Comment (al1img) [inline smclient/smclient.go:339]: Add empty line

File: smclient/smclient.go [hunk 22]
```diff
@@ -0,0 +1,822 @@
...
+	if err := client.stream.Send(&pb.SMOutgoingMessages{
+		SMOutgoingMessage: &pb.SMOutgoingMessages_UnitConfigStatus{UnitConfigStatus: status},
+	}); err != nil {
+		log.Errorf("Can't send unit config status: %v", err)
+	}
+}
+
+func (client *SMClient) processSetUnitConfig(config *pb.SetUnitConfig) {
+	status := &pb.UnitConfigStatus{}
+	if err := client.unitConfigProcessor.CheckUnitConfig(config.UnitConfig, config.VendorVersion); err != nil {
```
Comment (al1img) [inline smclient/smclient.go:352]: ditto

File: smclient/smclient_test.go [hunk 23]
```diff
@@ -0,0 +1,666 @@
...
+
+func TestSmRegistration(t *testing.T) {
+	server, err := newTestServer(serverURL)
+	if err != nil {
+		t.Fatalf("Can't create test server: %v", err)
+	}
+
+	defer server.close()
+
+	client, err := smclient.New(&config.Config{CMServerURL: serverURL}, "mainSM", "model1", nil, nil, nil, nil, nil, nil,
```
Comment (al1img) [inline smclient/smclient_test.go:112]: long line

File: smclient/smclient.go [hunk 24]
```diff
@@ -0,0 +1,822 @@
...
+const (
+	cmRequestTimeout   = 30 * time.Second
+	cmReconnectTimeout = 10 * time.Second
+)
+
+/***********************************************************************************************************************
+ * Vars
+ **********************************************************************************************************************/
+
+var errIncorrectAlertType = errors.New("incorrect alert type")
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply smclient/smclient.go:59]: done

File: smclient/smclient.go (hunk 19 above)
Comment (Eabramov84) [reply smclient/smclient.go:178]: done

File: smclient/smclient.go (hunk 20 above)
Comment (Eabramov84) [reply smclient/smclient.go:183]: fixed

File: smclient/smclient.go (hunk 21 above)
Comment (Eabramov84) [reply smclient/smclient.go:339]: done

**Review by al1img — APPROVED** [review]

**Pull request merged**